File: pt-2/614.second-degree-follower.java
/**
 * This LeetCode problem is a database problem, so the executable solution is
 * SQL rather than Java. The original Python file stores the SQL string; this
 * Java translation preserves that solution in a constant.
 *
 * Algorithm:
 * Self-join Follow where a followed user is also a follower. Group that middle
 * user and count distinct people they follow.
 *
 * Database complexity:
 * Dominated by the self-join and grouping. Indexes on follower and followee
 * are useful.
 */
class Solution {
    static final String SOLUTION_SQL = """
        SELECT f1.followee AS follower, COUNT(DISTINCT f2.followee) AS num
        FROM Follow f1
        JOIN Follow f2 ON f2.follower = f1.followee
        GROUP BY f1.followee
        ORDER BY f1.followee;
        """;
}

File: pt-2/614.second-degree-follower.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <random>
#include <regex>
#include <set>
#include <sstream>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

const char* SOLUTION_SQL = R"SQL(
SELECT f1.followee AS follower, COUNT(DISTINCT f2.followee) AS num
FROM Follow f1
JOIN Follow f2 ON f2.follower = f1.followee
GROUP BY f1.followee
ORDER BY f1.followee;
)SQL";

/*
Interview explanation:
A second-degree follower is a user who is followed and also follows other users. Self-join where f1.followee = f2.follower, then count distinct people they follow.

C++ data structures: none; SQL is embedded as documentation and answer.

Edge cases: inner join excludes users with no outgoing follows. DISTINCT avoids duplicate relationship counts.

Complexity: database self-join cost; indexes on follower and followee are useful.
*/
